### PR TITLE
[FX] Add delegate interface

### PR DESCRIPTION
**Pull request by jamesr66a**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#42865 [FX] Add delegate interface**

This adds an argument, `delegate` to `symbolic_trace` that allows the user to specify a delegate object to override certain behavior during symbolic tracing. Currently, the two behaviors that can be overridden are 1) leaf module identification and 2) the behavior of inserting a node. The latter can be used, for example, to exclude in-place operations from the trace.

Overriding `create_node` allows one to throw an exception when an erroneous condition is encountered. This exception will show a stack trace through the user's code indicating where the erroneous operation occured:

```
Traceback (most recent call last):
  File "test/test_fx.py", line 109, in test_disallow_override
    symbolic_trace(m2, delegate=NoMutableCallDelegate())
  File "/data/users/jamesreed/pytorch/torch/fx/symbolic_trace.py", line 108, in symbolic_trace
    graph.output(fn(*args))
  File "test/test_fx.py", line 105, in forward
    torch.log_(x)
  File "/data/users/jamesreed/pytorch/torch/fx/node.py", line 57, in __torch_function__
    self.graph, 'call_function', orig_method, args, kwargs, name=self.graph._name(orig_method.__name__))
  File "test/test_fx.py", line 88, in create_node
    raise RuntimeError('In-place operations are not supported')
RuntimeError: In-place operations are not supported

```

Differential Revision: [D23056584](https://our.internmc.facebook.com/intern/diff/D23056584)

### CODE REVIEW

File: torch/fx/symbolic_trace.py [hunk 1]
```diff
@@ -43,15 +44,40 @@ def _patch_function(fn, nargs):
...
+#                      via this parameter.
+#   - is_valid_call : A callable to validate whether a given target, args, and
+#                     kwargs should be recorded. This can be used, for example,
+#                     to error out when symbolic tracing encounters an in-place
+#                     operation. The passed-in callable should throw an exception
+#                     if it encounters a call that is not supported.
+def symbolic_trace(root : torch.nn.Module,
+                   is_leaf_module : Callable[[torch.nn.Module], bool] = is_leaf_module,
+                   is_valid_call : Callable[
+                       [Union[str, Callable], Tuple[Any]], Dict[str, Any]] = default_is_valid_call):
```
Comment (rdzhabarov) [inline torch/fx/symbolic_trace.py:72]: we'd probably need to provide some helpers for common scenarios, no in-place, no dict args, etc.

does not need to be in this PR.

**Review by rdzhabarov — APPROVED** [review]
👍  nice!

**Review by zdevito — APPROVED** [review]
Looks good. This suggests a follow-up refactor that separates Node from the tracing process that I will submit as a follow-up PR

File: test/test_fx.py [hunk 2]
```diff
@@ -73,5 +78,65 @@ def forward(self, d):
...
 
+    def test_disallow_override(self):
+        # Custom delegate to disallow in-place tensor operations
+        class NoMutableCallDelegate(DefaultDelegate):
+            def create_node(self, graph : Graph, kind : str, target : Union[str, Callable],
+                            args : Tuple[Any], kwargs : Dict[str, Any], name : Optional[str] = None) -> Node:
+                name = target if isinstance(target, str) else torch.typename(target)
+                if name[-1] == '_':
+                    raise RuntimeError('In-place operations are not supported')
+                return super().create_node(kind, target, args, kwargs, name)
```
Comment (dzhulgakov) [inline test/test_fx.py:89]: nit: it seems to miss `graph` argument, but still works because we throw an error first

File: torch/fx/node.py [hunk 3]
```diff
@@ -20,14 +20,15 @@ def __init__(self, graph, name, op, target, args, kwargs):
...
         return self.name
 
     def __getattr__(self, k):
         # note: not added to the graph yet, if this is a method call
         # we peephole optimize to the method invocation
-        return Attribute(self, k)
+        return Attribute(self, k, self.delegate)
 
     def __call__(self, *args, **kwargs):
         return self._create_node('call_method', '__call__', [self] + args, kwargs)
```
Comment (dzhulgakov) [inline torch/fx/node.py:34]: should this one also be using delegate.create_node?

**Pull request closed without merging**